internal/controller/jetstream_controller.go

```diff
@@ -0,0 +1,160 @@
+package controller
+
+import (
+	"fmt"
+	js "github.com/nats-io/nack/controllers/jetstream"
+	api "github.com/nats-io/nack/pkg/jetstream/apis/jetstream/v1beta2"
+	"github.com/nats-io/nats.go/jetstream"
+	v1 "k8s.io/api/core/v1"
+	"strings"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"time"
+)
+
+type connectionOptions struct {
+	Account string   `json:"account"`
+	Creds   string   `json:"creds"`
+	Nkey    string   `json:"nkey"`
+	Servers []string `json:"servers"`
+	TLS     api.TLS  `json:"tls"`
+}
+
+type JetStreamController interface {
+	client.Client
+
+	// ReadOnly returns true when no changes should be made by the controller.
+	ReadOnly() bool
+
+	// ValidNamespace ok if the controllers namespace restriction allows the given namespace.
+	ValidNamespace(string string) bool
+
+	// WithJetStreamClient provides a jetStream client to the given operation.
+	// The client uses the controllers connection configuration merged with opts.
+	//
+	// The given opts values take precedence over the controllers base configuration.
+	//
+	// Returns the error of the operation or errors during client setup.
+	WithJetStreamClient(opts *connectionOptions, op func(js jetstream.JetStream) error) error
+}
+
+func NewJSController(k8sClient client.Client, natsConfig *NatsConfig, controllerConfig *Config) (JetStreamController, error) {
+
+	return &jsController{
+		Client:           k8sClient,
+		config:           natsConfig,
+		controllerConfig: controllerConfig,
+	}, nil
+}
+
+type jsController struct {
+	client.Client
+	config           *NatsConfig
+	controllerConfig *Config
+}
+
+func (c *jsController) ReadOnly() bool {
+	return c.controllerConfig.ReadOnly
+}
+
+func (c *jsController) ValidNamespace(targetNamespace string) bool {
+	ns := c.controllerConfig.Namespace
+	return ns == "" || ns == targetNamespace
+}
+
+func (c *jsController) WithJetStreamClient(opts *connectionOptions, op func(js jetstream.JetStream) error) error {
+
+	// Build single use client
+	// TODO(future-feature): Use client-pool instead of single use client
+	cfg := c.buildNatsConfig(opts)
+
+	jsClient, closer, err := CreateJetStreamClient(cfg, true)
+	if err != nil {
+		return fmt.Errorf("create jetstream client: %w", err)
+	}
+	defer closer.Close()
+
+	return op(jsClient)
+}
+
+// buildNatsConfig uses given opts to override the base NatsConfig.
+func (c *jsController) buildNatsConfig(opts *connectionOptions) *NatsConfig {
+
+	serverUrls := strings.Join(opts.Servers, ",")
+
+	// Takes opts values if present
+	cfg := &NatsConfig{
+		CRDConnect: false,
+		ClientName: c.config.ClientName,
+		ServerURL:  or(serverUrls, c.config.ServerURL),
+		TLSFirst:   c.config.TLSFirst, // TODO(review): should this value depend on any opts? There is no TLSFirst in the spec
```

Comment:
If we have a `TLSFirst` option in the base config we should probably offer it in the spec as well, but at the moment, no there wouldn't be any spec option for it.